src/models/state/mod.rs

```diff
@@ -711,26 +722,36 @@ impl GlobalState {
         // monitored UTXO in the database. All monitored UTXOs are fetched outside
         // of the loop to avoid DB access/IO inside the loop.
         let mut recovery_data_for_missing_mutxos = vec![];
-        let mutxos = self.wallet_state.wallet_db.monitored_utxos().get_all();
-        '_outer: for incoming_utxo in incoming_utxos.into_iter() {
-            'inner: for monitored_utxo in mutxos.iter() {
-                if monitored_utxo.utxo == incoming_utxo.utxo {
-                    let msmp_res = monitored_utxo.get_latest_membership_proof_entry();
-                    let msmp = match msmp_res {
-                        Some((_blockh_hash, msmp_val)) => msmp_val,
-                        None => continue 'inner,
-                    };
-
-                    // If UTXO matches, then check if the AOCL index is also a match.
-                    // If it is, then the UTXO is already in the wallet database.
-                    if msmp.auth_path_aocl.leaf_index == incoming_utxo.aocl_index {
-                        continue '_outer;
+
```

Comment:
The previous version handled all DB IO outside of the loop (referring to line 735 in the new version). IIuc, this is no longer true. Isn't that a performance degradation?

Reply:
I will review.

Reply:
Analogy: This is basically the difference between reading a file in one call vs reading it line by line.  Reading the entire file in one call is fine if it is guaranteed to be a small file and/or you have low concurrency.   However, if you are reading in a 1 gigabyte file and you have much concurrency, your process will choke.  

Specifics:
Ok, in this case the previous code was calling monitored_utxos.get_all() before the loop.   This means it must read all available monitored_utxos into memory, all rows are collected into a Vec.  The collect itself takes some time.    I'm unclear how many utxos there could be, but I imagine that for some wallets (exchanges, frequent traders, multi-user wallets etc) it could be quite large with possibly tens of thousands of utxos.  So this could use significant amount of memory.   And the memory usage increases by the number of concurrent tasks performing same/similar operations.  Also, internally get_all() is reading records from the DB one-by-one anyway.  

The new code reads utxos one by one, so there is no need to allocate a bunch of memory up-front.  There is no collect() into a Vec.  We begin processing the first utxo right away.

In the past I benchmarked get_all (wraps get_many) vs get() in a loop.  iirc, get_all() was marginally faster but not enough to warrant increased memory, etc.   And it may be that it was only faster when it had pending writes, which meant that some items were in cache.   I would need to review the benches to be certain.

So yeah, mainly when I made this change I was thinking about avoiding the collect() and saving memory and generally being more concurrent.

related:  the new code is using an async `Stream` ([Stream primer](https://www.qovery.com/blog/a-guided-tour-of-streams-in-rust/)) instead of an iterator.  A `Stream` can potentially provide items in any order as each becomes available and the caller doesn't need to care.  At present, the storage layer that provides the `Stream` is not doing this, it calls get().await on each index in the order received.  But it could be modified to fire off all the gets and join_all() the results, so they come back in order of completion.  That *might* be faster, or it might create disk contention and be slower, I dunno.  Would need testing, and doesn't seem a priority now.

Reply:
The function in questions is invoked by the main thread through `global_state_lock.resync_memebership_proofs()` it holds a write lock over the global state until it's done, i.e. until all monitored UTXOs are in an updated state. So any concurrency would have to be in threads that don't need to read from the global state.

The same is true for `prune_abandoned_monitored_utxos` in this same file: This is done by the main thread, and all other threads pretty much have to wait for that to finish anyway.

The most critical point of operation for this software is when the global state is updated with a new block. The main thread grabs a write lock to the global state, and most operations (mining, sharing blocks with peers, answering RPC calls, creating new transactions) should not and cannot function until a new valid, consistent state is reached. And the main thread will communicate that this is the case by releasing its lock on global state.

This means that the most important things to aim for for this software is
1. Correctness and atomicity of state updates (the lock on global state achieves this)
2. Fast state updates, so other tasks can continue
3. Concurrency in the case where a write lock is *not* held on the global state.

**Reading all monitored UTXOs into memory**
I calculated how big a monitored UTXO is, assuming 1bn UTXOs, the number I expect after 5-10 years:
```
Size of `MonitoredUtxo`, in `BFieldElement`
UTXO: 14
  digest: 5
  coins: 9
    type_script_hash: 5
    state: 4
spent/confirmed/abandoned_at:  3*8 = 24
blockhash_to_membership_proof (3x): 47980
  digest: 5
  MmrMembershipProof: (~30 * 5) = 150
  MsMembershipProof:
    2x digest: 10
    ChunkDictionary (max 45 of these): 45 * (1 + 150 + 200) = 15795
    	u64: 1
    	MmrMembershipProof: (~30 * 5)
    	Chunk: (~0.10 * 2048) ~= 200
```

Rounding up, that's 50.000 `BFieldElement`s per monitored UTXO, which is 400.000 bytes, 400KB. Normal users will have between 0 and 100 monitored UTXOs which is 40MB. Loading that into memory is not a problem on any computer, not even a smartphone.

People running exchanges can have up to 10.000 monitored UTXOs, so around 4GB. If you're running an archival node in a setting that handles high volume, I think 4GB of RAM is absolutely within reason.

Reply:
Thank-you for the memory usage calculations.  I'm surprised at the 400Kb and 40Mb figures for 100 mutxo.  That's even higher than I expected.

Of course I agree we want fast state updates.  That is something we should always strive for.  So then, I do not see any reason to believe that fetching all records from the database in a for loop and collecting into a Vec (which is what get_all() does) and then reading them in a second loop would be faster than fetching each from the database in a single for loop and never collecting into a Vec, which is what the code in this PR does.  Indeed, on the face of it, it seems like this version should be faster.

Rather than guessing and theorizing, it is better to benchmark.   There is an existing benchmark for DbtVec I made last year.  It doesn't yet build for the new async storage, but I just ran the version from a week ago, which should be largely equivalent for comparing get() vs get_all().  Note that get_all() calls get_many() internally.  Here are the results:

```
db_dbtvec                      fastest       │ slowest       │ median        │ mean          │ samples │ iters
├─ read_100_entries                          │               │               │               │         │
│  ├─ get_each_entry_1_time                  │               │               │               │         │
│  │  ├─ get_many_persisted    88.17 µs      │ 159.2 µs      │ 116.8 µs      │ 107.4 µs      │ 100     │ 100
│  │  ├─ get_persisted         82.15 µs      │ 130.5 µs      │ 103.7 µs      │ 95.3 µs       │ 100     │ 100
│  │  ├─ get_many_unpersisted  20.47 µs      │ 26.95 µs      │ 20.95 µs      │ 21.1 µs       │ 100     │ 100
│  │  ╰─ get_unpersisted       10.83 µs      │ 11.54 µs      │ 10.91 µs      │ 10.92 µs      │ 100     │ 100
│  ╰─ get_each_entry_20_times                │               │               │               │         │
│     ├─ get_many_persisted    1.686 ms      │ 2.345 ms      │ 1.717 ms      │ 1.806 ms      │ 100     │ 100
│     ├─ get_persisted         1.495 ms      │ 1.682 ms      │ 1.541 ms      │ 1.542 ms      │ 100     │ 100
│     ├─ get_many_unpersisted  304.4 µs      │ 388.6 µs      │ 314.1 µs      │ 318.2 µs      │ 100     │ 100
│     ╰─ get_unpersisted       156.3 µs      │ 195.8 µs      │ 165.8 µs      │ 163.3 µs      │ 100     │ 100
```

So we see here that for 100 records it is faster to call get() directly than to call get_many().  This may be due to additional logic we have in get_many or it may be due largely to the collect().  I would need to re-code dbtvec get_many() to isolate that.  Anyway, it is clear that get_many() and by extension get_all() is not faster, but actually slower both for persisted and unpersisted vecs.

With regards to concurrency and memory usage:

In the case of  restore_monitored_utxos_from_recovery_data() that is operating on locked global state, it is correct that only one loop will be executing at a time.   However, if we look at for example get_latest_balance_height_internal() which also iterates over all mutxo, that only acquires a read lock.  This means that N number of operations can occur concurrently.  And that means that memory usage is linear with the number of concurrent operations.  So our busy exchange node using 4Gb ram per operation that may be processing many concurrent operations hits 32Gb with just 8 concurrent operations.  That is of course an extreme example, but it is useful to consider edge cases.

bottom line:   The code in this PR reflects my understanding of the most performant way to retrieve these DB values, and it is my recommendation that we leave it as-is.  In my opinion, get_all() would be allocating memory unnecessarily and be slower.  So I hope that puts your mind(s) at ease.   Nevertheless, I will change it to use get_all() for all these loops if you should ask for it again.

Reply:
It does put my mind somewhat at ease. Going forward, I think we should take two initiatives:
1. Measure and (log-info) output the time it takes to apply a new block to the archival state. I want to track that number and see how it scales with the number of MUTXOs.
2. Minimize duration of the write-lock. Our "scheduled tasks" (peer discovery, synchronization, mempool cleanup etc.) generally work such that they *ask* "is this action needed" and then only performs the action if this boolean evaluates to true. Resolving the boolean value should be done only with a read-lock on the global state.

Reply:
I created an issue for each of these.